connectors/flink/src/main/scala/io/delta/flink/internal/KernelSnapshotDelegator.scala

```diff
@@ -0,0 +1,99 @@
+package io.delta.standalone.internal
+
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.Path
+
+import io.delta.flink.source.internal.enumerator.supplier.KernelSnapshotWrapper
+import io.delta.kernel.internal.{SnapshotImpl => SnapshotImplKernel}
+import io.delta.standalone.DeltaScan
+import io.delta.standalone.actions.{AddFile => AddFileJ, Metadata => MetadataJ}
+import io.delta.standalone.data.{CloseableIterator, RowRecord => RowParquetRecordJ}
+import io.delta.standalone.expressions.Expression
+import io.delta.standalone.internal.actions.{AddFile, Metadata, Protocol, SetTransaction}
+import io.delta.standalone.internal.scan.DeltaScanImpl
+import io.delta.standalone.internal.util.ConversionUtils
+
+
+/**
+ * This class is designed to be passed to OptimisticTransactionImpl, and provide exactly what that
+ * needs to operate, but on a Kernel Snapshot rather than a standalone Snapshot.
+ *
+ * The methods/variables used by OptimisticTransactionImpl that we implement are:
+ *  - protocolScala
+ *  - metadataScala
+ *  - version
+ *  - getMetadata
+ *
+ *  Other functions that are used, we do not implement, but fall back to standalone. These functions
+ *  are only called in "exceptional" cases, so should not overly impact performance.
+ *  They include:
+ *  - scanScala (only called in markFilesAsRead, not used by flink)
+ *  - transactions (only used in txnVersion, which is used only on first commit for a flink app)
+ *    - This is a val, but it calls setTransactionsScala, so we log for that
+ *  - numOfFiles (only used in verifySchemaCompatibility, which happens only when a metadata update occures)
+ *  - allFilesScala (only used in verifySchemaCompatibility)
+ */
+class KernelSnapshotDelegator(
+    kernelSnapshot: SnapshotImplKernel,
+    hadoopConf: Configuration,
+    path: Path,
+    override val version: Long,
+    kernelDeltaLog: KernelDeltaLogDelegator,
+    standaloneDeltaLog: DeltaLogImpl)
+  extends SnapshotImpl(hadoopConf, path, -1, LogSegment.empty(path), -1, standaloneDeltaLog, -1) {
+
+  val snapshotWrapper = new KernelSnapshotWrapper(kernelSnapshot)
+  lazy val standaloneSnapshot: SnapshotImpl = standaloneDeltaLog.getSnapshotForVersionAsOf(getVersion())
+
+  /**
+   * Internal vals we need to override
+   */
+  override lazy val protocolScala: Protocol = {
+    val kernelProtocol = kernelSnapshot.getProtocol()
+    new Protocol(kernelProtocol.getMinReaderVersion(), kernelProtocol.getMinWriterVersion())
+  }
+
+  override lazy val metadataScala: Metadata = {
+    val metadata = snapshotWrapper.getMetadata()
+    ConversionUtils.convertMetadataJ(metadata)
+  }
+
+  // provide a path to use the faster txn lookup in kernel
+  def getLatestTransactionVersion(id: String): Option[Long] = {
+    val versionJOpt = kernelSnapshot.getLatestTransactionVersion(id)
+    if (versionJOpt.isPresent) {
```

Comment:
you can't just return `kernelSnapshot.getLatestTransactionVersion(id).asScala`? look for other usages of this in the code base? should be easy to convert java optional to scala option

Reply:
I don't think so. https://www.scala-lang.org/api/2.13.6/scala/jdk/OptionConverters$.html doesn't exist until scala 2.13